### ISSUE TITLE
Replace Redis with LevelDB

### ISSUE THREAD

**Issue by ekristen** [ekristen/docker-index#11]
Most people are running a single index and for 99% of private registry implementations a non-redundant storage system will be more then adequate. Also for the majority of private registries a single index server will be more then sufficient as well.

By replacing redis with LevelDB, its a 1 for 1 swap but allows the docker-index to only have the dependency of the registry docker image, and all data can be stored within the docker-index and/or use a host volume mount to expose the data for persistence.

**Comment by ekristen** [comment]
WIP: https://github.com/ekristen/docker-index/tree/level

**Comment by Caligatio** [comment]
If I submitted a pull request to make the storage backend modular, would you be interested?  I'm looking at Redis and any-db but, if done correctly, LevelDB could easily be added as well.

**Comment by ekristen** [comment]
@Caligatio I'm all for making the backend storage more modular, however I am not a fan of any-db, so I will not support a PR like that. However I do have a solution I've wanted to implement for a while, but I've had zero people ask for it, let alone let me know that they are using the project.

@Caligatio are you actively using the project? If so, what datastore would you prefer to be using?

Everything is setup for key/value storage, that is why `leveldb` is nice, because I can make the app entirely self-contained. However I've been wanting to move `app.js` to an exportable function like `index.createServer` where options could be passed. One of these options could be a datastore object.

That datastore object would have to contain 4 methods (`get`, `set`, `del`, and `expire`) From there how and where the data gets stored it wouldn't matter, any other solution would require a very large rewrite, that I am not sure I'd be willing to support at this time.

I might just take a stab at the createServer function this weekend and if I do I'll push it to a branch so you can test! 

I welcome all feedback, comments and other that you may have one way or the other.

**Comment by ekristen** [comment]
So the basic premise of what I'd like to do will work, except I'll need to pull over elements from my leveldb branch because of things like `smembers`.

**Comment by Caligatio** [comment]
I actually decided on Friday to try to code a Docker Index implementation in Node but then found that you already done so.  If things go well during the test, I think I have a decent user base (100s) but the authentication/authorization part has to be rock solid.

I'm fairly flexible on backends but it *needs* to be reliable.  any-db provided the options of Postgres and MySQL (and presumably MariaDB) which is why it was attractive.  I could just as easily support MongoDB.  Any reason you dislike any-db?  I'm not saying have that be the modular option, just one of many options to include Redis and LevelDB.

My idea was to create a backend object that exposed standard user-related functions (createUser, getPermissions, etc) and then the backend object would have to translate into the backend speak.  I think this would be easier, more flexible, and cleaner than trying to expose backend primitives like set, get, etc.

**Comment by ekristen** [comment]
I've been using my implementation for over a year.  I have dozens of users, servers and CI systems that publish and pull daily if not hourly from it without issue. 

The authentication works just like the public docker index.

What would help me is understanding what is your reasoning for wanting to use Postgres or MySQL for the backed?

I like simplicity. Using something like any-db you force yourself into a corner, have to use their methods and their supported backends. Granted I forced myself into a corner with redis originally but I've had some philosophical changes in the way I code since then and just haven't had a need to upgrade the project to match. Mainly RDMS on something like is is extremely overkill. 

Key/value storage is far simpler and you get just as good of stability if not even better, so many less moving pieces than a RDMS.

Id like for you to use the project but I have unit tests and the current code is stable and working I'd be hesitant to make the sweeping changes that an any-db would require. 

I'm all for talking though it a bit more.

**Comment by ekristen** [comment]
@Caligatio so I was just giving this a little more thought. Trying to remove my personal feelings a bit, hard to do to be honest ... :) 

I'm all for supporting external data sources, however I'm not ok with any-db, or forcing this into having to use a `SQL` language or `RDMS` they are overly complex, and I'm all for simplicity. 

If we can figure out a way to make it so that you or anyone could easily drop in a datastore object that exposed certain functions that the application needs and that is it, I'm all for it.

At the same time I'd like to prevent the situation of you or someone else forking this and releasing yet another docker-index that just would confuse people. Docker has an amazing ecosystem, competition is healthy, most indexes that exist exist to serve a purpose of a product, I've tried to release something that just works for everyone and isn't tied to a specific product.

My core requirements is that the default datastore to be leveldb (it is redis right now, but I've already made steps to make it the primary, just haven't released them yet)

If you'd like to discuss offline I'm all for it, just email me, we can do an IRC chat or something. I'm all for abstracting the data store out a bit, just want to do it in a smart way.
